GVFS/GVFS.Upgrader/UpgradeOrchestrator.cs

```diff
@@ -0,0 +1,389 @@
+        {
+            this.upgrader = upgrader;
+            this.tracer = tracer;
+            this.preRunChecker = preRunChecker;
+            this.output = output;
+            this.input = input;
+            this.remount = false;
+            this.shouldExitOnError = shouldExitOnError;
+            this.ExitCode = ReturnCode.Success;
+        }
+
+        public UpgradeOrchestrator()
+        {
+            string logFilePath = GVFSEnlistment.GetNewGVFSLogFileName(
+                ProductUpgrader.GetLogDirectoryPath(),
+                GVFSConstants.LogFileTypes.UpgradeProcess);
+            JsonTracer jsonTracer = new JsonTracer(GVFSConstants.GVFSEtwProviderName, "UpgradeProcess");
+            jsonTracer.AddLogFileEventListener(
+                logFilePath,
+                DefaultEventLevel,
+                Keywords.Any);
+
+            this.tracer = jsonTracer;
+            this.preRunChecker = new InstallerPreRunChecker(this.tracer);
+            this.upgrader = new ProductUpgrader(ProcessHelper.GetCurrentProcessVersion(), this.tracer);
+            this.output = Console.Out;
+            this.input = Console.In;
+            this.remount = false;
+            this.shouldExitOnError = false;
+            this.ExitCode = ReturnCode.Success;
+        }
+
+        public ReturnCode ExitCode { get; private set; }
+
+        public void Execute()
+        {
+            string error = null;
+
+            if (this.upgrader.IsNoneRing())
+            {
+                this.output.WriteLine(GVFSConstants.UpgradeVerbMessages.NoneRingConsoleAlert);
+                this.output.WriteLine(GVFSConstants.UpgradeVerbMessages.SetUpgradeRingCommand);
+            }
+            else
+            {
+                try
+                {
+                    Version newVersion = null;
+                    if (!this.TryRunUpgradeInstall(out newVersion, out error))
+                    {
+                        this.ExitCode = ReturnCode.GenericError;
+                    }
+                }
+                finally
+                {
+                    string remountError = null;
+                    if (!this.TryRemountRepositories(out remountError))
+                    {
+                        remountError = Environment.NewLine + "WARNING: " + remountError;
+                        this.output.WriteLine(remountError);
+                        this.ExitCode = ReturnCode.Success;
+                    }
+
+                    this.DeletedDownloadedAssets();
+                }
+            }
+            
+            if (this.ExitCode == ReturnCode.GenericError)
+            {
+                error = Environment.NewLine + "ERROR: " + error;
+                this.output.WriteLine(error);
+            }
+
+            if (this.input == Console.In)
+            {
+                this.output.WriteLine("Press Enter to exit.");
+                this.input.ReadLine();
+            }
+            
+            if (this.shouldExitOnError)

```

Comment:
Can you drop this bool if you just set `Environment.ExitCode` instead of calling `Environment.Exit`?

Reply:
Yes, that is possible. Removed the bool.